segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -230,6 +236,17 @@ private boolean isShutdownException(Throwable ex) {
         return result;
     }
 
+    /**
+     * Gets the maximum number of Operations to fetch from the operation queue. This is calculated based on the estimated
+     * cache insertion capacity and its goal is to reduce the number of operations we have in flight as we near the
+     * maximum configured cache capacity.
+     *
+     * @return The maximum number of Operations to fetch from the operation queue.
+     */
+    private int getFetchCount() {
+        return Math.max(1, (int) (this.cacheUtilizationProvider.getCacheInsertionCapacity() * MAX_READ_AT_ONCE));
```

Comment:
While I understand that the cache throttler needs to take into account in-flight data, I have a couple of questions about dynamically changing the size of the set of elements picked for processing:
1) I may be probably wrong, but my understanding is that picking elements from this in-memory queue is so fast, that it may not be so different in terms of processing time doing 10 iterations of 100 elements than 100 iterations of 10 elements... However, I see that it is useful to control the amount of in-flight data during saturation periods.
2) The items picked from that queue are aggregated anyway in a DataFrame to be written to Tier 1, right? In this sense, is this indirectly doing some kind of throttling on Tier1 as well?

Reply:
In general, picking items from the queue is very fast. However, if we start to throttle, we won't be picking items from that queue that frequently and it will start to grow. In addition, once we pick items from the queue we cannot put them back. We either have to process the operation (validate it, write it to BK, put it in cache, etc), or cancel it. We have no intention of cancelling it at this point since that will mess things up for the upstream code. In addition, once we pick them up, these operations will spend a significant amount of time waiting to be acked from BK (millis or even hundreds of millis for busy systems) - there is nothing we can do to abort these operations at this point.

The cache-based throttler used to make its decision based on the data that already existed in the cache, however it did not account for data that we have already committed to processing (in-flight) which will make it to the cache at any time. It is for this reason that I added this extra piece of information. Once the cache nears capacity, picking a large number of operations to commit (at once) risks pushing us beyond that capacity and thus into "Cache Full" territory. However, if we reduce the amount of data that we pick at once we can fine-tune this and thus make sure we won't run into that situation. 

As an analogy, think how you would fill a glass of water without spilling over and without waiting an eternity - while it's still mosty empty, you fill it using a maximum flow, but as it gets full, you reduce the flow so you can properly stop when it's full without spilling water over. In this example, the glass of water is the cache and the faucet is the Operation Processor and the flow of water is managed using throttling and picking the number of items to process at once.

> is this indirectly doing some kind of throttling on Tier1 as well

Yes. All throttling is done before Tier 1.